public/components/trace_analytics/components/common/shared_components/custom_datagrid.tsx

```diff
@@ -137,14 +137,38 @@ export const RenderCustomDataGrid: React.FC<RenderCustomDataGridParams> = ({
       )
     : [];
 
+  const isLastPage =
+    tracesTableMode === 'traces' && pagination?.pageSize && maxTraces != null
+      ? (pagination.pageIndex + 1) * pagination.pageSize >= maxTraces &&
+        rowCount > maxTraces &&
+        maxTraces < MAX_DISPLAY_ROWS
+      : false;
+
+  // Used for when trace count less than default maxTraces
+  useEffect(() => {
+    if (
+      tracesTableMode === 'traces' &&
+      rowCount > 0 &&
+      rowCount < maxTraces &&
+      maxTraces !== rowCount
+    ) {
+      setMaxTraces(rowCount);
+    }
+  }, [rowCount, maxTraces, tracesTableMode]);
+
   useInjectElementsIntoGrid(
     rowCount,
     MAX_DISPLAY_ROWS,
     tracesTableMode ?? '',
     () => {
-      setMaxTraces((prevMax: number) => Math.min(prevMax + 500, MAX_DISPLAY_ROWS));
+      // Add 500 more traces at a time until all traces present
+      const nextIncrement = Math.min(500, rowCount - maxTraces);
```

Comment:
may be we can declare this function first and then add it as a param to the `useInjectElementsIntoGrid` function

Reply:
Updated to
```
  const incrementTraceCount = () => {
    const nextIncrement = Math.min(500, rowCount - maxTraces);
    if (nextIncrement > 0 && maxTraces < MAX_DISPLAY_ROWS) {
      setMaxTraces((prevMax) => Math.min(prevMax + nextIncrement, MAX_DISPLAY_ROWS));
    }
  };

  useInjectElementsIntoGrid(
    rowCount,
    MAX_DISPLAY_ROWS,
    tracesTableMode ?? '',
    incrementTraceCount,
    maxTraces,
    isLastPage
  );
  ```